exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,245 @@
+    ///
+    /// # Notes
+    ///
+    /// The default implementation returns `null`. For transactions defined with
+    /// the [`message!`] macro, you may redefine `info()` as
+    ///
+    /// ```
+    /// # #[macro_use] extern crate exonum;
+    /// extern crate serde_json;
+    /// # use exonum::blockchain::Transaction;
+    /// # use exonum::blockchain::transaction::ExecutionContext;
+    /// # use exonum::storage::Fork;
+    ///
+    /// message! {
+    ///     struct MyTransaction {
+    ///         // Transaction definition...
+    /// #       const TYPE = 1;
+    /// #       const ID = 1;
+    /// #       const SIZE = 8;
+    /// #       field foo: u64 [0 => 8]
+    ///     }
+    /// }
+    ///
+    /// impl Transaction for MyTransaction {
+    ///     // Other methods...
+    /// #   fn verify(&self) -> bool { true }
+    /// #   fn execute(&self, _: &mut ExecutionContext) {}
+    ///
+    ///     fn info(&self) -> serde_json::Value {
+    ///         serde_json::to_value(self).expect("Cannot serialize transaction to JSON")
+    ///     }
+    /// }
+    /// # fn main() { }
+    /// ```
+    ///
+    /// [`TxInfo.content`]: ../explorer/struct.TxInfo.html#structfield.content
+    /// [explorer]: ../explorer/index.html
+    /// [`message!`]: ../macro.message.html
+    fn info(&self) -> Value {
+        Value::Null
+    }
+}
+
+/// Execution status of the transaction.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExecutionStatus {
+    /// Successful transaction execution.
+    Succeeded,
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// General failure (unspecified reason).
+    Failed,
+    /// User defined execution status. Can have different meanings for different transactions and
+    /// services.
+    Custom(u8),
+}
+
+/// `Transaction`'s execution context.
+#[derive(Debug)]
+pub struct ExecutionContext<'a> {
+    fork: &'a mut Fork,
+    status: Option<ExecutionStatus>,
+    changes_number: usize,
+}
+
+impl<'a> ExecutionContext<'a> {
+    /// Creates a new `ExecutionContext` instance.
+    pub fn new(fork: &'a mut Fork) -> Self {
+        Self {
+            changes_number: fork.patch().len(),
+            fork,
+            status: None,
+        }
+    }
+
+    /// Returns execution status consuming `ExecutionContext`.
+    pub fn into_status(self) -> ExecutionStatus {
+        match self.status {
+            Some(val) => val,
+            None => {
+                if self.fork.patch().len() > self.changes_number {
+                    ExecutionStatus::Succeeded
+                } else {
+                    ExecutionStatus::Failed
+                }
+            }
+        }
+    }
+
+    /// Returns reference to the `Fork`.
+    pub fn fork(&mut self) -> &mut Fork {
+        self.fork
+    }
+
+    /// Sets `Transaction`'s execution status. See `ExecutionStatus` for the details.
+    pub fn set_execution_status(&mut self, status: ExecutionStatus) {
```

Comment:
Is it possible to set `Panic` status? If yes, I'd say this is a suboptimal design and may warrant separate types for:

- Execution status that can be set by the developer, and
- Execution status that is persisted to the storage